---
title: "Analysis of single-cell RNA-seq data using a topic model, Part 2: practical implementation"
author: Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{Analysis of single-cell RNA-seq data using a topic model, Part 2: practical implementation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette continues from [Part 1][vignette-part-1], where we
introduced the basic concepts behind an analysis of single-cell
RNA-seq data using a topic model. Here we revisit the same single-cell
data set, giving more detailed guidance, and discussing a few 
complications that may arise.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

We begin our analysis by loading the packages. Then we set the seed so
that the results can be reproduced.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Matrix)
library(fastTopics)
library(ggplot2)
library(cowplot)
set.seed(1)
```

Now we load the data set and the $K = 6$ pre-fitted topic model.

```{r load-data}
data(pbmc_4k)
counts <- pbmc_4k$counts
fit <- pbmc_4k$fit
```

Assessing model fit
-------------------

The topic model was fitted using the `fit_topic_model` function. This
function, as we mentioned, hides most of the complexities of model
fitting. Nevertheless, it is a good idea to check that the model
fitting has converged to a maximum-likelihood solution. This can be
easily done in fastTopics:

```{r plot-loglik, fig.height=2, fig.width=4}
fit_nmf <- multinom2poisson(fit)
plot_progress_poisson_nmf(fit_nmf,x = "iter",add.point.every = 10,
                          colors = "black") +
  theme_cowplot(font_size = 10)
```

Internally, `fit_topic_model` first fits a non-negative matrix
factorization to the count data, then converts the fitted NMF to a
topic model, so the plot shows the improvement in the NMF model fit
over time (improvement is measured by the NMF likelihood, not the
multinomial topic model likelihood). Judging by this plot, the
parameter estimates get close to a maxium-likelihood solution after
about 150 updates.

For larger or more complex data sets, more than 200 updates may be
needed to obtain a good fit. For guidance on fine-tuning the model
fitting, refer to the documentation for `fit_topic_model` and
`fit_poisson_nmf`.

Evaluating fit: single cell likelihoods
---------------------------------------

Another feature of the topic model is that it provides a likelihood
for each cell which quantifies how well the topic model "fits" the
observed expression:

```{r loglik-1}
loglik <- loglik_multinom_topic_model(counts,fit)
```

This could be useful for comparing different topic models, say, topic
models with different settings of $K$. It could also be used to
isolate "outlier" cells---that is, cells with expression that is
poorly captured by the topic model. In the PBMC data, a histogram of
the log-likelihoods shows that the topic model provides a very small
fit for a small number of cells:

```{r loglik-2, fig.height=2, fig.width=3.5, eval=FALSE}
p4 <- ggplot(data.frame(x = loglik),aes(x)) +
  geom_histogram(color = "black",fill = "black",bins = 64) +
  theme_cowplot(font_size = 10) +
  labs(x = "loglik",y = "number of cells")
print(p4)
```

Most of the outliers are in the CD34+ subpopulation:

```{r loglik-3, eval=FALSE}
i <- which(loglik < -5000)
table(samples[i,"subpop"])
```

Visualizing the topics without cell labels
------------------------------------------

In Part 1, we made use of additional information about the cells to
help us interpret the topics. In some cases, however, we may not have
the benefit of external information to guide our explorations. In this
section, we illustrate the *de novo* discovery of structure using a
topic model.

Without the cell labels, the Structure plot can still be quite
effective:

```{r structure-plot-without-labels, fig.width=6.5, fig.height=1.5, results="hide", message=FALSE}
topic_colors <- c("skyblue","forestgreen","darkmagenta","dodgerblue",
                  "gold","darkorange")
p1 <- structure_plot(fit,topics = 1:6,colors = topic_colors)
print(p1)
```

From this Structure plot, we see the clustering of the cells into B
cells (dark blue), CD14+ monocytes (green) and CD34+ cells (purple).
The clustering of the NK (light blue) and T cells (mixtures of yellow
and orange) is less distinct, and this is reflected by more sharing of
topics (mostly topic 6) among NK and T cells.

Of course, without the cell labels, we cannot know that the topics
correspond to these cell types without further downstream
analyses---for this, we will perform a differential expression
analysis (continue reading).

Identifying clusters from the mixture proportions
-------------------------------------------------

In more complex data sets, some types of fine scale structure, or rare
cell types, may not show up well in the Structure plot, at least
without additional fine-tuning of the plot.

One simple strategy we have found to be generally helpful is to first
identify clusters in the mixture proportions matrix, $L$. Here, we
illustrate this idea using *k*-means. Specifically, we apply $k$-means
to the principal components (PCs) of the mixture proportions.

We previously experimented with different numbers of clusters and we
found that 6 clusters worked best. Note that *k*-means can be
sensitive to (random) initialization, so we set the seed before
running *k*-means to ensure that the output is reproducible.

```{r clustering-1}
set.seed(1)
pca <- prcomp(fit$L)$x
clusters <- kmeans(pca,centers = 6,iter.max = 100)$cluster
summary(factor(clusters))
```

Now we create another Structure plot in which the cells are organized
by their membership to the 6 clusters:

```{r structure-plot-by-cluster-1, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
p2 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = factor(clusters),gap = 25)
print(p2)
```

Judging by the Structure plot, the *k*-means clusters look sensible,
the one exception being that the T cells (cells with high proportions
of topics 5 and 6) seem to be rather arbitrarily split into clusters 1
and 3. Therefore, we merge these two clusters:

```{r clustering-2}
clusters[clusters == 3] <- 1
clusters <- factor(clusters)
```

We have found that visual inspection of the clusters followed by
manual refinement is often be needed to get the "right" clustering.

With the benefit of hindsight, we know that these 5 clusters
correspond to 5 cell types: B cells, T cells, natural killer (NK)
cellls, CD14+ monoytes and CD34+ cells. For clarity of presentation,
we label the clusters by these cell types, noting that, in most
practical settings, labeling the clusters is usually not possible
until downstream analyses have been performed.

```{r structure-plot-by-cluster-2, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
levels(clusters) <- c("T","CD14+","B","CD34+","NK")
p3 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = clusters,gap = 25)
print(p3)
```

This new Structure plot better highlights the heterogeneity within the
clusters, particularly in the T cells. Within the T cells cluster,
there is distinctive subset of cells with contributions from topic 1
(the "NK cells" topic, in light blue).

For clustering, it is also sometimes helpful to visualize the clusters
in PCA plots which show the projection of the cells onto PCs of the
mixture proportions:

```{r pca-plot-1, fig.width=3.5, fig.height=2.5, message=FALSE}
cluster_colors <- c("gold","forestgreen","dodgerblue","darkmagenta","skyblue")
p4 <- pca_plot(fit,fill = clusters) +
  scale_fill_manual(values = cluster_colors)
print(p4)
```

(Note that the CD14+ and CD34+ cells separate better in PCs 3 and 4.)

When there are many overlapping points, as is the case here, a
plotting instead the density of the points can show the clustering
more clearly:

```{r pca-plot-2, fig.width=3.75, fig.height=2.5, message=FALSE}
p5 <- pca_hexbin_plot(fit,bins = 24)
print(p5)
```

In particular, the separation of the T cells and NK cells into two
clusters is more clear in the density ("hexbin") plot.

This example is only meant to illustrate clustering of the cells by
the estimated mixture proportions; *k*-means may not be the best
clustering technique for all data sets. For more complex data sets,
some ingenuity may be needed to identify and visualize clusters.

More on differential expression analysis
----------------------------------------

In Part 1, we performed a differential expression analysis using the
topic model, and using the results of this analysis we were able to
interpret topic 1 as NK cells and topic 4 as B cells. Here, we use
perform differential expression analysis to tackle interpretation of
the topics that capturing variation in the T cells cluster.

Unlike the B cells and NK cells, expression in T cells is captured by
mixtures of two topics. To isolate T-cell-specific expression
patterns, we perform differential expression analysis on the combined
topic:

```{r volcano-plot-t, fig.height=3.5, fig.width=4.5, results="hide"}
genes <- pbmc_4k$genes
fit_merged <- merge_topics(fit,5:6)
diff_count_merged <- diff_count_analysis(fit_merged,counts)
p6 <- volcano_plot(diff_count_merged,k = "k5+k6",label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 16)
print(p6)
```

In the volcano plot, the T-cell marker gene *CD3D* is one of the top
genes.

When the volcano plot shows many overlapping differentially expressed
genes, like it does here, it can be helpful to explore the results
interactively. The function `volcano_plotly` creates an interactive
volcano plot that can be viewed in a Web browser:

```{r volcano-plotly}
p7 <- volcano_plotly(diff_count_t,k = 5,file = "volcano_plot_t_cells.html",
                     labels = genes$symbol)
```

This call creates a new file `volcano_plot_t_cells.html`. The
interactive volcano plot can also be viewed [here][volcano-plotly].

Within the T cell cluster, there is a distinctive subset of cells with
sizeable contributions from the "NK cells" topic (that is, topic 1,
shown in light blue in the Structure plot). This subset of cells sits
in between the dense clumps of NK and T cells in the PCA plots
above. To highlight this subset of cells---which turn out to be CD8+
cells, as we will see shortly---we define a new cluster:

```{r structure-plot-by-cluster-3, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
clusters <- as.character(clusters)
clusters[clusters == "T" & fit$L[,1] > 0.1] <- "CD8+"
clusters <- factor(clusters)
p8 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = factor(clusters),gap = 25)
print(p8)
```

Now we can perform differential expression analysis on the CD8+
cluster (as well as the other clusters) using the
`diff_count_clusters` function, and visualize the results using a
volcano plot:

```{r volcano-plot-cd8, fig.height=3.5, fig.width=4.5, results="hide"}
dfc_out <- diff_count_clusters(counts,clusters)
p9 <- volcano_plot(dfc_out,k = "CD8+",label_above_quantile = 0.995,
                   labels = genes$symbol,max.overlaps = 10) +
  ggtitle("cluster CD8+")
print(p9)
```

For example, *CD8A* is one of the top differentially expressed genes,
a clue that this cluster corresponds to CD8+ cytotoxic T cells.

Topics 5 and 6 capture continuous heterogeneity within the T cells. To
isolate the differences in expression these two topics, we perform a
differential expression analysis on the T cells only:

```{r volcano-plot-cd4, fig.height=3.5, fig.width=4.5, results="hide"}
i <- which(clusters == "T")
fit_t <- select_loadings(fit,loadings = i)
diff_count_t <- diff_count_analysis(fit_t,counts[i,])
p8 <- volcano_plot(diff_count_t,k = 5,label_above_quantile = 0.995,
                   labels = genes$symbol,max.overlaps = 20)
print(p8)
```

Judging by top differentially expressed genes such as *CD4* and
*CCR10*, topic 5 likely corresponds to CD4+/CD8+ lineage
differentiation in T cells.

Session info
------------

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```

[vignette-part-1]: https://stephenslab.github.io/fastTopics/articles/single_cell_rnaseq_basic.html
[volcano-plotly]: https://stephenslab.github.io/fastTopics/articles/volcano_plot_t_cells.html
[zheng-2017]: https://doi.org/10.1038/ncomms14049
