<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Mixture of 10 FACS-purified PBMC Single-Cell RNA-seq data — pbmc_facs • fastTopics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Mixture of 10 FACS-purified PBMC Single-Cell RNA-seq data — pbmc_facs"><meta property="og:description" content="These data are a selection of the reference
  transcriptome profiles generated via single-cell RNA sequencing
  (RNA-seq) of 10 bead-enriched subpopulations of PBMCs (Donor A),
  described in Zheng et al (2017). The data are unique
  molecular identifier (UMI) counts for 16,791 genes in 3,774 cells.
  (Genes with no expression in any of the cells were removed.) Since
  the majority of the UMI counts are zero, they are efficiently
  stored as a 3,774 x 16,791 sparse matrix. These data are used in
  the vignette illustrating how 'fastTopics' can be used to analyze to
  single-cell RNA-seq data. Data for a separate set of 1,000 cells is
  provided as a &amp;#8220;test set&amp;#8221; to evaluate out-of-sample predictions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6-147</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stephenslab/fastTopics" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Mixture of 10 FACS-purified PBMC Single-Cell RNA-seq data</h1>
    <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/HEAD/R/pbmc_facs.R" class="external-link"><code>R/pbmc_facs.R</code></a></small>
    <div class="hidden name"><code>pbmc_facs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These data are a selection of the reference
  transcriptome profiles generated via single-cell RNA sequencing
  (RNA-seq) of 10 bead-enriched subpopulations of PBMCs (Donor A),
  described in Zheng <em>et al</em> (2017). The data are unique
  molecular identifier (UMI) counts for 16,791 genes in 3,774 cells.
  (Genes with no expression in any of the cells were removed.) Since
  the majority of the UMI counts are zero, they are efficiently
  stored as a 3,774 x 16,791 sparse matrix. These data are used in
  the vignette illustrating how 'fastTopics' can be used to analyze to
  single-cell RNA-seq data. Data for a separate set of 1,000 cells is
  provided as a “test set” to evaluate out-of-sample predictions.</p>
    </div>


    <div id="format">
    <h2>Format</h2>
    <p><code>pbmc_facs</code> is a list with the following elements:</p>
<dl><dt>counts</dt>
<dd><p>3,774 x 16,791 sparse matrix of UMI counts, with
     rows corresponding to samples (cells) and columns corresponding to
     genes. It is an object of class <code>"dgCMatrix"</code>).</p></dd>


  <dt>counts_test</dt>
<dd><p>UMI counts for an additional test set of 100
    cells.</p></dd>


  <dt>samples</dt>
<dd><p>Data frame containing information about the
    samples, including cell barcode and source FACS population
    (“celltype” and “facs_subpop”).</p></dd>


  <dt>samples_test</dt>
<dd><p>Sample information for the additional test
     set of 100 cells.</p></dd>


  <dt>genes</dt>
<dd><p>Data frame containing information and the genes,
    including gene symbol and Ensembl identifier.</p></dd>


  <dt>fit</dt>
<dd><p>Poisson non-negative matrix factorization (NMF) fitted
    to the UMI count data <code>counts</code>, with rank <code>k = 6</code>. See
    the vignette how the Poisson NMF model fitting was performed.</p></dd>


  <dt>de</dt>
<dd><p>Result of calling 
    <code>de_analysis(fit,counts,pseudocount = 0.1,
    control = list(ns = 1e4,nc = 4))</code>
    after first setting the seed to 1, <code>set.seed(1)</code>.</p></dd>

</dl></div>
    <div id="source">
    <h2>Source</h2>
    <p><a href="https://www.10xgenomics.com/resources/datasets" class="external-link">https://www.10xgenomics.com/resources/datasets</a></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>G. X. Y. Zheng <em>et al</em> (2017). Massively parallel digital
transcriptional profiling of single cells. <em>Nature Communications</em>
<b>8</b>, 14049. doi: <a href="https://doi.org/10.1038/ncomms14049" class="external-link">10.1038/ncomms14049</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Number of cells: %d\n"</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of cells: 3774</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Number of genes: %d\n"</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of genes: 16791</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Proportion of counts that are non-zero: %0.1f%%.\n"</span>,</span>
<span class="r-in">            <span class="fl">100</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Proportion of counts that are non-zero: 4.3%.</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew
Stephens.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

